src/js/links.js

```diff
@@ -0,0 +1,251 @@
+const LINKS_GET_LINKS_LIMIT = 100;
+
+const Links = {
+
+  maxNumLinks: LINKS_GET_LINKS_LIMIT,
+
+  /**
+   * A mapping from each provider to an object { sortedLinks, siteMap, linkMap }.
+   * sortedLinks is the cached, sorted array of links for the provider.
+   * siteMap is a mapping from base domains to URL count associated with the domain.
+   *         The count does not include blocked URLs. siteMap is used to look up a
+   *         user's top sites that can be targeted with a suggested tile.
+   * linkMap is a Map from link URLs to link objects.
+   */
+  _providers: new Map(),
+
+  /**
+   * A list of objects that are observing links updates.
+   */
+  _observers: [],
+
+  /**
+   * The properties of link objects used to sort them.
+   */
+  _sortProperties: [
+    "frecency",
+    "lastVisitDate",
+    "url",
+  ],
+
+  /**
+   * Compares two links.
+   *
+   * @param {Link} aLink1 The first link.
+   * @param {Link} aLink2 The second link.
+   * @return {Integer} A negative number if aLink1 is ordered before aLink2, zero if
+   *         aLink1 and aLink2 have the same ordering, or a positive number if
+   *         aLink1 is ordered after aLink2.
+   *
+   * NOTE: compareLinks's this object is bound to Links below.
+   */
+  compareLinks(aLink1, aLink2) {
+    for (var prop of this._sortProperties) {
+      if (!(prop in aLink1) || !(prop in aLink2)) {
+        throw new Error("Comparable link missing required property: " + prop);
+      }
+    }
+    return aLink2.frecency - aLink1.frecency ||
+           aLink2.lastVisitDate - aLink1.lastVisitDate ||
+           aLink1.url.localeCompare(aLink2.url);
+  },
+
+  /**
+   * Merges the cached lists of links from all providers whose lists are cached.
+   *
+   * @return {Array} The merged list.
+   */
+  _getMergedProviderLinks() {
+    // Build a list containing a copy of each provider's sortedLinks list.
+    var linkLists = [];
+    for (var provider of this._providers.keys()) {
+      /*if (!AllPages.enhanced && provider != PlacesProvider) {
+        // Only show history tiles if we're not in 'enhanced' mode.
+        continue;
+      }*/
+      var links = this._providers.get(provider);
+      if (links && links.sortedLinks) {
+        linkLists.push(links.sortedLinks.slice());
+      }
+    }
+
+    function getNextLink() {
+      var minLinks = null;
+      for (var links of linkLists) {
+        if (links.length &&
+            (!minLinks || Links.compareLinks(links[0], minLinks[0]) < 0)) {
+          minLinks = links;
+        }
+      }
+      return minLinks ? minLinks.shift() : null;
+    }
+
+    var finalLinks = [];
+    for (var nextLink = getNextLink();
+         nextLink && finalLinks.length < this.maxNumLinks;
+         nextLink = getNextLink()) {
+      finalLinks.push(nextLink);
+    }
+
+    return finalLinks;
+  },
+
+  /**
+   * Gets the current set of links contained in the grid.
+   *
+   * @return {Object} The links in the grid.
+   */
+  getLinks() {
```

Comment:
Note: I wanted to include your use of `reduce()` using `fillGaps()`. However, `fillGaps()` returns filler which isn't what we'd want to concatenate our remaining links to (even if we used links instead of sites). Unless I'm missing something there?